src/credential/mod.rs

```diff
@@ -1,45 +1,159 @@
+};
+use sd_jwt::{SdJwt, SdJwtError};
+use serde::{Deserialize, Serialize};
+use ssi::prelude::AnyJsonCredential;
 
 /// An unparsed credential, retrieved from storage.
-#[derive(Debug, Serialize, Deserialize, uniffi::Record)]
+#[derive(Debug, Clone, Serialize, Deserialize, uniffi::Object)]
 pub struct Credential {
     /// The local ID of this credential.
-    pub id: Uuid,
+    pub(crate) id: Uuid,
     /// The format of this credential.
-    pub format: CredentialFormat,
+    pub(crate) format: CredentialFormat,
     /// The type of this credential.
-    pub r#type: CredentialType,
+    pub(crate) r#type: CredentialType,
     /// The raw payload of this credential. The encoding depends on the format.
-    pub payload: Vec<u8>,
+    pub(crate) payload: Vec<u8>,
     /// The alias of the key that is authorized to present this credential.
-    pub key_alias: Option<KeyAlias>,
+    pub(crate) key_alias: Option<KeyAlias>,
+}
+
+#[uniffi::export]
+impl Credential {
+    /// Create a new credential.
+    #[uniffi::constructor]
+    pub fn new(
+        id: Uuid,
+        format: CredentialFormat,
+        r#type: CredentialType,
+        payload: Vec<u8>,
+        key_alias: Option<KeyAlias>,
+    ) -> Arc<Self> {
+        Arc::new(Self {
+            id,
+            format,
+            r#type,
+            payload,
+            key_alias,
+        })
+    }
+
+    /// Return the ID of the credential.
+    pub fn id(&self) -> Uuid {
+        self.id
+    }
+
+    /// Return the format of the credential.
+    pub fn format(&self) -> CredentialFormat {
+        self.format.clone()
+    }
+
+    /// Return the type of the credential.
+    pub fn r#type(&self) -> CredentialType {
+        self.r#type.clone()
+    }
+
+    /// Return the raw payload of the credential.
+    pub fn payload(&self) -> Vec<u8> {
+        self.payload.clone()
+    }
+}
+
+// Internal helper methods.
+impl Credential {
+    /// Convert the parsed credential into a specialized JSON credential.
+    pub fn try_into_parsed(&self) -> Result<Arc<ParsedCredential>, CredentialDecodingError> {
+        self.to_owned().try_into()
+    }
+
+    /// Returns the credential as a `Mdoc` credential.
+    pub fn try_into_mdoc(&self) -> Result<Arc<Mdoc>, CredentialDecodingError> {
+        self.to_owned()
+            .try_into()
+            .map_err(|e: MdocInitError| CredentialDecodingError::MsoMdoc(e.to_string()))
+    }
+
+    /// Returns the credential as a `JwtVc` credential.
+    pub fn try_into_jwt_vc(&self) -> Result<Arc<JwtVc>, CredentialDecodingError> {
+        self.to_owned()
+            .try_into()
+            .map_err(|e: JwtVcInitError| CredentialDecodingError::JwtVc(e.to_string()))
+    }
+
+    /// Return the credential as an AnyJsonCredential type.
+    pub fn try_into_any_json_credential(
+        &self,
+    ) -> Result<AnyJsonCredential, CredentialDecodingError> {
+        self.to_owned().try_into()
+    }
+}
```

Comment:
This is not necessary, we should just go through `ParsedCredential` for consistency.

Reply:
Yes, we can remove these methods.

I've kept `try_into_parsed` method, but removed the others.